File: sdk/runanywhere-flutter/lib/core/protocols/lifecycle/model_lifecycle_protocol.dart
// Model lifecycle state and management protocols
// Matches iOS ModelLifecycleProtocol from Core/Protocols/Lifecycle/ModelLifecycleProtocol.swift

/// Represents the various states in a model's lifecycle
enum ModelLifecycleState {
  uninitialized('uninitialized'),
  discovered('discovered'),
  downloading('downloading'),
  downloaded('downloaded'),
  extracting('extracting'),
  extracted('extracted'),
  validating('validating'),
  validated('validated'),
  initializing('initializing'),
  initialized('initialized'),
  loading('loading'),
  loaded('loaded'),
  ready('ready'),
  executing('executing'),
  error('error'),
  cleanup('cleanup');

  final String rawValue;

  const ModelLifecycleState(this.rawValue);

  /// Whether the model is currently processing an operation
  bool get isProcessing {
    switch (this) {
      case ModelLifecycleState.downloading:
      case ModelLifecycleState.extracting:
      case ModelLifecycleState.validating:
      case ModelLifecycleState.initializing:
      case ModelLifecycleState.loading:
      case ModelLifecycleState.executing:
        return true;
      default:
        return false;
    }
  }
}

/// Observer protocol for model lifecycle changes
abstract class ModelLifecycleObserver {
  /// Called when the model transitions to a new state
  void modelDidTransition(
    ModelLifecycleState oldState,
    ModelLifecycleState newState,
  );

  /// Called when an error occurs during state transition
  void modelDidEncounterError(Object error, ModelLifecycleState state);
}

/// Protocol for managing model lifecycle states
abstract class ModelLifecycleManager {
  /// Current state of the model
  ModelLifecycleState get currentState;

  /// Transition to a new state
  Future<void> transitionTo(ModelLifecycleState state);

  /// Add an observer for state changes
  void addObserver(ModelLifecycleObserver observer);

  /// Remove an observer
  void removeObserver(ModelLifecycleObserver observer);

  /// Check if a transition is valid
  bool isValidTransition(ModelLifecycleState from, ModelLifecycleState to);
}

/// Errors related to model lifecycle
sealed class ModelLifecycleError implements Exception {
  String get message;
}

class InvalidTransitionError extends ModelLifecycleError {
  final ModelLifecycleState from;
  final ModelLifecycleState to;

  InvalidTransitionError(this.from, this.to);

  @override
  String get message =>
      'Invalid transition from ${from.rawValue} to ${to.rawValue}';

  @override
  String toString() => 'InvalidTransitionError: $message';
}

class StatePrerequisiteNotMetError extends ModelLifecycleError {
  final String reason;

  StatePrerequisiteNotMetError(this.reason);

  @override
  String get message => 'State prerequisite not met: $reason';

  @override
  String toString() => 'StatePrerequisiteNotMetError: $message';
}

class TransitionFailedError extends ModelLifecycleError {
  final Object cause;

  TransitionFailedError(this.cause);

  @override
  String get message => 'Transition failed: $cause';

  @override
  String toString() => 'TransitionFailedError: $message';
}

class InvalidStateError extends ModelLifecycleError {
  final String reason;

  InvalidStateError(this.reason);

  @override
  String get message => 'Invalid state: $reason';

  @override
  String toString() => 'InvalidStateError: $message';
}

File: sdk/runanywhere-react-native/src/Core/Protocols/Lifecycle/ModelLifecycleProtocol.ts
/**
 * ModelLifecycleProtocol.ts
 *
 * Protocol for managing model lifecycle states
 *
 * Reference: sdk/runanywhere-swift/Sources/RunAnywhere/Core/Protocols/Lifecycle/ModelLifecycleProtocol.swift
 */

/**
 * Represents the various states in a model's lifecycle
 */
export enum ModelLifecycleState {
  Uninitialized = 'uninitialized',
  Discovered = 'discovered',
  Downloading = 'downloading',
  Downloaded = 'downloaded',
  Extracting = 'extracting',
  Extracted = 'extracted',
  Validating = 'validating',
  Validated = 'validated',
  Initializing = 'initializing',
  Initialized = 'initialized',
  Loading = 'loading',
  Loaded = 'loaded',
  Ready = 'ready',
  Executing = 'executing',
  Error = 'error',
  Cleanup = 'cleanup',
}

/**
 * Whether the model is currently processing an operation
 */
export function isModelLifecycleProcessing(state: ModelLifecycleState): boolean {
  return (
    state === ModelLifecycleState.Downloading ||
    state === ModelLifecycleState.Extracting ||
    state === ModelLifecycleState.Validating ||
    state === ModelLifecycleState.Initializing ||
    state === ModelLifecycleState.Loading ||
    state === ModelLifecycleState.Executing
  );
}

/**
 * Observer protocol for model lifecycle changes
 */
export interface ModelLifecycleObserver {
  /**
   * Called when the model transitions to a new state
   * @param oldState - Previous state
   * @param newState - New state
   */
  modelDidTransition(from: ModelLifecycleState, to: ModelLifecycleState): void;

  /**
   * Called when an error occurs during state transition
   * @param error - The error that occurred
   * @param state - The state where the error occurred
   */
  modelDidEncounterError(error: Error, inState: ModelLifecycleState): void;
}

/**
 * Protocol for managing model lifecycle states
 */
export interface ModelLifecycleManager {
  /**
   * Current state of the model
   */
  readonly currentState: ModelLifecycleState;

  /**
   * Transition to a new state
   * @param state - The target state
   * @throws Error if the transition is invalid
   */
  transitionTo(state: ModelLifecycleState): Promise<void>;

  /**
   * Add an observer for state changes
   * @param observer - The observer to add
   */
  addObserver(observer: ModelLifecycleObserver): void;

  /**
   * Remove an observer
   * @param observer - The observer to remove
   */
  removeObserver(observer: ModelLifecycleObserver): void;

  /**
   * Check if a transition is valid
   * @param from - Source state
   * @param to - Target state
   * @returns Whether the transition is valid
   */
  isValidTransition(from: ModelLifecycleState, to: ModelLifecycleState): boolean;
}

/**
 * Errors related to model lifecycle
 */
export class ModelLifecycleError extends Error {
  constructor(
    message: string,
    public readonly fromState?: ModelLifecycleState,
    public readonly toState?: ModelLifecycleState,
    public readonly reason?: string
  ) {
    super(message);
    this.name = 'ModelLifecycleError';
  }

  static invalidTransition(
    from: ModelLifecycleState,
    to: ModelLifecycleState
  ): ModelLifecycleError {
    return new ModelLifecycleError(
      `Invalid transition from ${from} to ${to}`,
      from,
      to
    );
  }

  static statePrerequisiteNotMet(reason: string): ModelLifecycleError {
    return new ModelLifecycleError(`State prerequisite not met: ${reason}`, undefined, undefined, reason);
  }

  static transitionFailed(error: Error): ModelLifecycleError {
    return new ModelLifecycleError(
      `Transition failed: ${error.message}`,
      undefined,
      undefined,
      error.message
    );
  }

  static invalidState(reason: string): ModelLifecycleError {
    return new ModelLifecycleError(`Invalid state: ${reason}`, undefined, undefined, reason);
  }
}
